README.md
# NAME

WWW::Myki - A simple Perl interface to Myki online account management portal

# VERSION 0.01

# SYNOPSIS

    use WWW::Myki;

    my $myki = WWW::Myki->new(
				username => 'myusername',
				password => 'mypassw0rd'
			);

    # Print card number, card holder, Myki money and Myki pass balances
   

    foreach $card ( $myki->cards ) {
      print "Card number: ". $card->id ." - Holder: ". $card->holder "\n";
      print "Myki money balance : ". $card->money ." - Myki pass balance: ". $card->pass ."\n";
    }

    # Print the date, time, service, description and cost of our last 15 transactions

    foreach my $trip ( $card->transactions ) { 
      printf( "%10s %8s %-10s %-20s %-5s\n", $trip->date, $trip->time, 
              $trip->service, $trip->desc, $trip->debit )
    } 
      

# DESCRIPTION

[WWW::Myki](http://search.cpan.org/perldoc?WWW::Myki) provides a simple interface to the Myki online account management portal functionality
for registered Myki users.

# METHODS

## new ( %args )

Constructor.  Creates a new WWW::Myki object and  attempts a login using the provided credentials.
On successful login, returns a WWW::Myki object.

The constructor accepts an anonymous hash of two mandatory parameters:

- username

Your registered Myki account username.

- password

Your registered Myki account password.

## cards

  my @cards = $myki->cards;

Returns an array of [WWW::Myki::Card](http://search.cpan.org/perldoc?WWW::Myki::Card) objects.  Each object is representative of a Myki card
registered to the specified account.

# AUTHOR

Luke Poskitt, `<ltp at cpan.org>`

# BUGS

Please report any bugs or feature requests to `bug-www-myki at rt.cpan.org`, or through
the web interface at [http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Myki](http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Myki).  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

# SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Myki

You can also look for information at:

- RT: CPAN's request tracker

[http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Myki](http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Myki)

- AnnoCPAN: Annotated CPAN documentation

[http://annocpan.org/dist/WWW-Myki](http://annocpan.org/dist/WWW-Myki)

- CPAN Ratings

[http://cpanratings.perl.org/d/WWW-Myki](http://cpanratings.perl.org/d/WWW-Myki)

- Search CPAN

[http://search.cpan.org/dist/WWW-Myki/](http://search.cpan.org/dist/WWW-Myki/)

# LICENSE AND COPYRIGHT

Copyright 2012 Luke Poskitt.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

# SEE ALSO

[WWW::Myki::Card](http://search.cpan.org/perldoc?WWW::Myki::Card), [WWW::Myki::Transaction](http://search.cpan.org/perldoc?WWW::Myki::Transaction)


Changes
Revision history for WWW-Myki

0.01    Fri 5 Oct 2012 ltp@cpan.org
        Initial release.

0.02    Sat 6 Oct 2012 ltp@cpan.org
	Fixed Makefile.PL


lib/WWW/Myki/Card.pm
package WWW::Myki::Card;

use strict;
use warnings;

use HTML::TreeBuilder;
use Scalar::Util qw(weaken);
use Carp qw(croak);

our $VERSION	= '0.01';
our @MATTR	= qw(id _link holder money pass);
our @ATTR	= qw(id _link holder money pass type expiry status money_top_up
		     money_total active_pass inactive_pass last_transaction);

foreach my $attr ( @ATTR ) {
	{    
		no strict 'refs';
		*{ __PACKAGE__ .'::'. $attr } = sub { 
			my( $self, $val ) = @_;
			$self->{$attr} = $val if defined $val;
			return $self->{$attr} 
		}
	}   
}

sub new {
        my( $class, %args ) = @_; 
        my $self = {}; 
        bless $self, $class;
	$args{_mech} ? weaken( $self->{_mech} = $args{_mech} ) : croak "Mandatory attribute _mech not supplied";

        foreach my $attr ( @MATTR ) { 
                $args{$attr} ? $self->{$attr} = $args{$attr} : croak "Mandatory attribute $attr not present";
        }   

	$self->_get_card_details;
        return $self
}

sub refresh {
	my $self = shift;
	$self->_get_card_details
}

sub _get_card_details {
        my $self = shift;
	my $r = $self->{_mech}->get( $self->{_link} );
	my $t = HTML::TreeBuilder->new_from_content( $r->content );
	$t = $t->look_down( id => 'ctl00_uxContentPlaceHolder_uxCardDetailsPnl' );
	my @r = $t->look_down ( _tag => 'td' );
        $self->type		( $r[3]->as_text  );
	$self->expiry		( $r[5]->as_text  );
        $self->status		( $r[7]->as_text  );
        $self->money_top_up	( $r[11]->as_text );
        $self->money_total	( $r[13]->as_text );
        $self->active_pass	( $r[15]->as_text );
        $self->inactive_pass	( $r[17]->as_text );
        $self->last_transaction	( $r[19]->as_text );
}

sub transactions {
        my( $self, $refresh ) = shift;
	return @{ $self->{transactions} } if ( $self->{transactions} and not $refresh );
	undef $self->{transactions};
	$self->{_mech}->get( $self->{_link} );
	$self->{_mech}->follow_link( text => 'My transactions' );
	$self->{_mech}->form_name( 'aspnetForm' );
	$self->{_mech}->select( 'ctl00$uxContentPlaceHolder$uxCardList', $self->id );
	my $r = $self->{_mech}->click_button( name => 'ctl00$uxContentPlaceHolder$uxSelectNewCard' );
	my $t = HTML::TreeBuilder->new_from_content( $r->content );
	( $self->{total_transactions} ) 
		= ( $t->look_down( id => 'ctl00_uxContentPlaceHolder_uxTotalRecords' )->as_text =~ /of\s(\d*)\s/ );
	$t = $t->look_down( id => 'ctl00_uxContentPlaceHolder_uxMykiTxnHistory' );

	for ( $t->look_down( _tag => 'tr' ) ) { 
		my @cols = map { $_->as_text } $_->look_down( _tag => 'td' ) or next;
		push @{ $self->{transactions} },  
			WWW::Myki::Transaction->new(
					date    => $cols[0],
					time    => $cols[1],
					type    => $cols[2],
					service => $cols[3],
					zone    => $cols[4],
					desc    => $cols[5],
					credit  => $cols[6],
					debit   => $cols[7],
					balance => $cols[8]
				)   
	}   

        return @{ $self->{transactions} }
}

=head1 NAME

WWW::Myki::Card - Class for operations with a Myki card

=head1 VERSION 0.01

=head1 SYNOPSIS

    # Print my Myki card money balance
    my $balance = $card->money;

    # What time did I stumble home last night?
    print $card->last_transaction;

    # Yeesh, really?  How?
    print $card->service;

    # Ooooohh, now I remember.
    
=head1 DESCRIPTION

L<WWW::Myki::Card> is a class providing account and card management and querying functionality
for registered Myki users.

Please note that you're are not meant to call the constructor yourself, instead a WWW::Myki::Card
object will be created automatically for you by calls to methods in a L<WWW::Myki> object like
B<cards>.

=head1 METHODS

=head2 id

Returns the card ID number.

=head2 holder

Returns the name of the registered card holder.

=head2 money

Returns the balance of Myki money on the card.

=head2 pass

Returns the balance of the active Myki pass (if any) on the card.

=head2 type

Returns the card type.

=head2 expiry

Returns the card expiry date in the format DD Mon YYYY, where Mon is the abbreviated month name.

=head2 status

Returns the card status.

=head2 money_top_up

Returns the value of any Myki money top up in progress - this is money that has been added by a top up, 
but may not yet have been credited to the card.

=head2 money_total

Returns the total of the balance of Myki money and the balance of Myki money top in progress on the card.

=head2 active_pass

Returns the balance of the current active Myki pass (if any) on the card.

=head2 inactive_pass

Returns the balance of the current inactive Myki pass (if any) on the card.

=head2 last_transaction

Returns the last transaction time and date for the card in the format; DD Mon YYYY HH:MM:SS AM/PM, 
where Mon is the abbreviated month name.

=head2 transactions

  foreach my $trip ( $card->transactions ) {
    printf( "%10s %8s %-10s %-20s\n", $trip->date, $trip->time, $trip->service, $trip->desc )
  }

  # Prints a formatted list of the last 15 transactions for this card - e.g.
  # 
  # 29/05/2012 17:28:38 Bus        Surburbia,Route SUB16out_new
  # 29/05/2012 08:08:12 Bus        Metro,Route MET16in_new

Returns an array of L<WWW::Myki::Transaction> objects representing the last 15 transactions for the card.

See L<WWW::Myki::Transaction> for more information on transactions.  Transaction data is cached on the initial
invocation to increase the performance of subsequent calls and reduce unnecessary communication with the Myki
portal.  This is probably what you want, but if you really do want to force transaction data to be refreshed
then you can call the method with the argument B<refresh> set to a true value. e.g.

  $card->transactions( refresh => 1 );

Please note that this will incur a performance penalty.

=head2 refresh

When a WWW::Myki::Card object is created, the card data is cached to improve the performance of subsequent
method calls and reduce unessecary network communication.  This is probably what you want, however if you
do want to force the object to update its cached data for any reason, then you can call B<refresh>. Note 
that doing so will incur a performance penalty.

=head1 AUTHOR

Luke Poskitt, C<< <ltp at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-myki-card at rt.cpan.org>, or through the web interface 
at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Myki-Card>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Myki::Card

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Myki-Card>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Myki-Card>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Myki-Card>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Myki-Card/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Luke Poskitt.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 SEE ALSO

L<WWW::Myki>, L<WWW::Myki::Transaction>

=cut

1;


lib/WWW/Myki/Transaction.pm
#!/usr/bin/perl

package WWW::Myki::Transaction;

use strict;
use warnings;

use Carp qw(croak);

our @ATTR = qw(credit zone time date service debit balance desc type);

sub new {
	my ($class, %args) = @_;
	my $self = {};
	bless $self, $class;

	foreach my $attr ( @ATTR ) {
		$args{$attr} ? $self->{$attr} = $args{$attr} : croak "Mandatory attribute $attr not present";
	}
	
	return $self
}

foreach my $attr (@ATTR) {
	{	
		no strict 'refs';
		*{ __PACKAGE__ .'::'. $attr } = sub { return $_[0]->{$attr} }
	}
}

=head1 NAME

WWW::Myki::Transaction - Class for operations with a Myki transaction

=head1 VERSION 0.01

=head1 SYNOPSIS

  # Print the date, time, service, description and cost of our last 15 transactions

  foreach my $trip ( $card->transactions ) { 
    printf( "%10s %8s %-10s %-20s %-5s\n", $trip->date, $trip->time, 
            $trip->service, $trip->desc, $trip->debit )
  } 

  # Maybe do a rough calculation of the cost of public transport per day

  use List::Util qw( sum );
  my %sum;

  foreach my $t ( $card->transactions ) {
    ( my $cost = $t->debit ) =~ s/(-|\$|\s)//g;
    $sum{ $t->date } += $cost;
  }

  print "Average daily cost of travel on public transport: ",
		( ( sum values %sum ) / ( keys %sum ) ), "\n";

  # Still cheaper than environmental cost of driving to work
    
=head1 DESCRIPTION

L<WWW::Myki::Transaction> is a class providing Myki transaction querying functionality for registered Myki users.
Each WWW::Myki::Transaction object is representative of a single Myki card transaction.

Please note that you're are not meant to call the constructor yourself, instead a WWW::Myki::Transaction
object will be created automatically for you by calls to methods in a L<WWW::Myki::Card> object like
B<transactions>.

=head1 METHODS

=head2 date

Return the date of the transaction using the format DD/MM/YYYY.

=head2 time

Print the time of the last transaction using the format HH:MM:SS AM/PM.

=head2 service

The type of service for the last transaction (e.g. busi or train)

=head2 zone

The transit zone for the last transaction.

=head2 desc

A description of the transaction usually including the start and destination localities and a service name.
This may be a hyphen (-) where no information is provided.

=head2 debit

The amount debited for the transaction using in decimal currency format (i.e. $D.DD) where present, or a hyphen
where no debit was made.

=head2 credit

The amount credited for the transaction using in decimal currency format (i.e. $D.DD) where present, or a hyphen
where no credit was made.

=head2 balance

The balance of the associated Myki card at the time this transaction was completed in decimal curreny format.

=head2 type

The type of transaction made (e.g. Touch-on or Touch-off).

=head1 AUTHOR

Luke Poskitt, C<< <ltp at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-myki-transaction at rt.cpan.org>, or through the web interface 
at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Myki-Transaction>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Myki::Transaction

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Myki-Transaction>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Myki-Transaction>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Myki-Transaction>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Myki-Transaction/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Luke Poskitt.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 SEE ALSO

L<WWW::Myki>, L<WWW::Myki::Card>

=cut


1;


lib/WWW/Myki.pm
package WWW::Myki;

use warnings;
use strict;

use WWW::Myki::Transaction;
use WWW::Myki::Card;
use WWW::Mechanize;
use HTML::TreeBuilder;
use Carp qw(croak);

our $VERSION = '0.03';

sub new {
	my( $class, %args ) = @_;
	my $self	= {};
	bless $self, $class;
	$args{username}	? $self->{username} = $args{username} : croak 'Constructor failed: must specify username';
	$args{password}	? $self->{password} = $args{password} : croak 'Constructor failed: must specify password';
	$self->{_mech}	= WWW::Mechanize->new();
	$self->{_url}	= {
			base	=> 'https://www.mymyki.com.au/NTSWebPortal/',
			account	=> 'Registered/MyMykiAccount.aspx?menu=My+myki+account',
			login	=> 'login.aspx'
			};
	$self->_login;
	return $self
}

sub _login {
	my $self = shift;
	$self->{_mech}->get( $self->{_url}->{base} . $self->{_url}->{login} );
	$self->{_mech}->form_name( 'aspnetForm' );
	$self->{_mech}->field( 'ctl00$uxContentPlaceHolder$uxUsername', $self->{username} );
	$self->{_mech}->field( 'ctl00$uxContentPlaceHolder$uxPassword', $self->{password} );
	$self->{_mech}->click( 'ctl00$uxContentPlaceHolder$uxLogin' );
	my %form = (
			'__EVENTTARGET'		=> 'ctl00$uxContentPlaceHolder$uxTimer',
			'__EVENTVALIDATION'	=> $self->{_mech}->value('__EVENTVALIDATION'),
			'__VIEWSTATE'		=> $self->{_mech}->value('__VIEWSTATE'),
			'ctl00$ScriptManager1'	=> 'ctl00$uxContentPlaceHolder$Panel|ctl00$uxContentPlaceHolder$uxTimer'
		);
	my $r 	= $self->{_mech}->post( $self->{_url}->{base} . $self->{_url}->{account}, \%form );
	my $t 	= HTML::TreeBuilder->new_from_content( $r->content );
	$t	= $t->look_down( id => 'ctl00_uxContentPlaceHolder_uxMyCards' );
	my @cards = $t->look_down( _tag => 'td' );
	@cards or croak 'Unable to retrieve card information';
	
	for( my $i=0; $i<@cards; $i+=4 ) {
		$self->{_cards}->{$cards[$i]->as_text} = 
			WWW::Myki::Card->new(
				id	=> $cards[$i]->as_text,
				_mech	=> $self->{_mech},
				_link	=> $self->{_url}->{base} . 'Registered/' . $cards[$i]->look_down( _tag => 'a' )->attr( 'href' ),
				holder	=> $cards[$i+1]->as_text,
				money	=> $cards[$i+2]->as_text,
				pass	=> $cards[$i+2]->as_text
			)
	}
}

sub cards { 
	return values %{$_[0]->{_cards}} 
}

=head1 NAME

WWW::Myki - A simple Perl interface to Myki online account management portal

=head1 VERSION 0.01

=head1 SYNOPSIS

    use WWW::Myki;

    my $myki = WWW::Myki->new(
				username => 'myusername',
				password => 'mypassw0rd'
			);

    # Print card number, card holder, Myki money and Myki pass balances
   
    foreach $card ( $myki->cards ) {
      print "Card number: ". $card->id ." - Holder: ". $card->holder "\n";
      print "Myki money balance : ". $card->money ." - Myki pass balance: ". $card->pass ."\n";
    }

    # Print the date, time, service, description and cost of our last 15 transactions

    foreach my $trip ( $card->transactions ) { 
      printf( "%10s %8s %-10s %-20s %-5s\n", $trip->date, $trip->time, 
              $trip->service, $trip->desc, $trip->debit )
    } 
      
=head1 DESCRIPTION

L<WWW::Myki> provides a simple interface to the Myki online account management portal functionality
for registered Myki users.

=head1 METHODS

=head2 new ( %args )

Constructor.  Creates a new WWW::Myki object and  attempts a login using the provided credentials.
On successful login, returns a WWW::Myki object.

The constructor accepts an anonymous hash of two mandatory parameters:

=over

=item *

username

Your registered Myki account username.

=item *

password

Your registered Myki account password.

=back

=cut

=head2 cards

  my @cards = $myki->cards;

Returns an array of L<WWW::Myki::Card> objects.  Each object is representative of a Myki card
registered to the specified account.

=head1 AUTHOR

Luke Poskitt, C<< <ltp at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-myki at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Myki>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Myki

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Myki>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Myki>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Myki>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Myki/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Luke Poskitt.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 SEE ALSO

L<WWW::Myki::Card>, L<WWW::Myki::Transaction>

=cut

1;
